Enable course search for the unified course experience

andy-armstrong opened the pull request:
## [LEARNER-1112](https://openedx.atlassian.net/browse/LEARNER-1112)

### Description

This PR upgrades course search to have its own page that is integrated into the new course experience. In addition, the code has been refactored into the openedx/feature directory.

![image](https://cloud.githubusercontent.com/assets/5985072/26459174/9a52f58c-4143-11e7-86c1-48f17cc53516.png)
...
![image](https://cloud.githubusercontent.com/assets/5985072/26459193/a495571a-4143-11e7-9965-7e738c0599ad.png)

### Sandbox

I was unable to set up course search on a sandbox so let me know if you'd like to see this locally. For now, see the screen shots above.

### Testing
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible
- [ ] safecommit violation code review process

Front-end changes:
- [ ] i18n
- [ ] Accessibility (Check for a11y violations, ensure keyboard accessible, screen reader testing as appropriate)
- [ ] RTL

### Reviewers
- [ ] Assign reviewers to your PR based on the changes it contains (Dev, Doc, UX, Accessibility, Product, DevOps)

List optional/FYI reviewers here:
 
### Post-review
- [ ] Rebase and squash commits

robrap commented on openedx/features/course_search/static/course_search/js/views/search_results_view.js:
```diff
@@ -75,31 +82,33 @@
-            // Bookmarks share the same container element, So we are doing
-            // this to ensure that elements are in clean/initial state
+                // Empty any loading/error message and empty the el
+                // Bookmarks share the same container element, So we are doing
```
I don't fully understand this comment.  Is it still accurate regarding 'Bookmarks'?

robrap commented on openedx/features/course_search/templates/course_search/course-search-fragment.html:
```diff
@@ -0,0 +1,70 @@
+                <div class="has-breadcrumbs">
+                    <div class="breadcrumbs">
+                        <span class="nav-item">
+                            <a href="${course_url}">Course</a>
```
Is this just placeholder text that doesn't need to be translated?

andy-armstrong replied:
Good catch. This should absolutely be translated!

robrap commented on openedx/features/course_search/views/course_search.py:
```diff
@@ -0,0 +1,66 @@
+    The home page for a course.
+    """
+    @method_decorator(login_required)
+    @method_decorator(ensure_csrf_cookie)
```
Nit: I don't think `ensure_csrf_cookie` is needed.

andy-armstrong replied:
I have to admit to not knowing when it is needed. :-) Why would it not be needed here?

robrap commented:
FYI: I've reviewed all changes through `a92e5c66fc4de7049463fca50e308f1711ba87cd`.

andy-armstrong commented:
@lizcohen This is the story to enable course search

dianakhuang commented:
jenkins run bokchoy

andy-armstrong commented:
@HarryRein it would be great if you could review this PR, particularly from a front-end perspective. Thanks!

andy-armstrong commented on lms/static/karma_lms.conf.js:
```diff
@@ -28,6 +28,7 @@ var options = {
     sourceFiles: [
         {pattern: 'coffee/src/**/!(*spec).js'},
         {pattern: 'course_bookmarks/**/!(*spec).js'},
+        {pattern: 'course_search/**/!(*spec).js'},
```
Good question, but no. The course experience JavaScript is bundled by Webpack automatically.

andy-armstrong commented on lms/static/lms/js/spec/main.js:
```diff
@@ -676,6 +676,7 @@
         'course_bookmarks/js/spec/bookmark_button_view_spec.js',
         'course_bookmarks/js/spec/bookmarks_list_view_spec.js',
         'course_bookmarks/js/spec/course_bookmarks_factory_spec.js',
+        'course_search/js/spec/course_search_spec.js',
```
This is a RequireJS configuration file, so course experience isn't listed because it is built via Webpack.

andy-armstrong commented on lms/templates/ux/reference/course-skeleton.html:
```diff
@@ -56,7 +56,10 @@ <h2 class="hd hd-2 page-title">Skeleton Page</h2>
+                        <button type="button" class="action action-clear aria-label="Clear search">
+                            <span class="icon fa fa-times-circle" aria-hidden="true"></span>
+                        </button>
+                        <button class="btn btn-small search-button" type="button">Search</button>
```
Yes. Buttons that are of type submit will submit the form when pressed.

andy-armstrong commented on openedx/features/course_search/static/course_search/js/spec/course_search_spec.js:
```diff
@@ -277,27 +276,6 @@ define([
-                beforeEach(function() {
-                    beforeEachHelper.call(this, CourseSearchItemView);
-                });
-                it('renders items correctly', rendersItem);
```
Thanks for checking. I removed these subclasses in the code, so the tests are no longer necessary.

andy-armstrong commented on openedx/features/course_search/static/course_search/js/spec/course_search_spec.js:
```diff
@@ -497,18 +449,20 @@ define([
                 it('shows loading message', showsLoadingMessage);
                 it('shows error message', showsErrorMessage);
-                it('returns to content', returnsToContent);
+                xit('returns to content', returnsToContent);
```
I spent a couple of days getting nowhere on why they have started failing. This code will go away as soon as the merged course/home tab goes out, so I'm tempted to just delete them .

andy-armstrong commented on openedx/features/course_search/static/course_search/js/views/search_item_view.js:
```diff
@@ -1,40 +1,43 @@
             attributes: {
-                'role': 'region',
+                role: 'region',
                 'aria-label': 'search result'
```
It is necessary because in JavaScript that would be evaluated as an expression.

andy-armstrong commented on common/test/acceptance/tests/lms/test_lms_cohorted_courseware_search.py:
```diff
@@ -110,6 +110,16 @@ def _goto_staff_page(self):
+        """
+        self.course_home_page.visit()
+        self.course_home_page.search_for_term(term)
+        course_search_results_page = CourseSearchResultsPage(self.browser, self.course_id).wait_for_page()
```
@HarryRein could you address @robrap's suggestions for the Bok Choy tests, since I think you wrote this (or was it @dianakhuang?)

HarryRein replied:
Yes, for sure, I'll take a look - this test might have been Diana but I wrote the accessibility tests that all work with the same general flow.

andy-armstrong commented:
@robrap thanks for the feedback. I've addressed all of your comments except for the Bok Choy ones which were not mine. Do you have a sense when you'll be able to complete the review (or did you complete it?).

dianakhuang commented:
jenkins run bokchoy

dianakhuang commented:
@robrap @HarryRein this should be ready for review.

dianakhuang commented:
jenkins run python

dianakhuang commented:
jenkins run python
